### PR TITLE
[BEAM-1726] TestFlinkRunner should assert PAssert success/failure counters

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aljoscha** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Rebased on top of master.
There is still one ROS test failing in Flink Runner currently.

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aljoscha** [comment]
This looks good to me now!

Might be good to have another pair of eyes on this, though.

R: @tgroh if you could have a look, that would be great.

**Comment by aviemzur** [comment]
Retest this please

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java [hunk 1]
```diff
@@ -163,11 +164,6 @@ public void addSideInputValue(
   @Override
   public <T> T get(PCollectionView<T> sideInput, BoundedWindow window) {
 
-    if (!isReady(sideInput, window)) {
```
Comment (tgroh) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java]: Why is this removed? Elements should not be processed until all of their side inputs are ready.

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResult.java [hunk 2]
```diff
@@ -53,13 +55,28 @@ public State getState() {
   @Override
   public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
       throws AggregatorRetrievalException {
+    return getAggregatorValues(aggregator.getName());
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T> T getAggregatorValue(String name) throws AggregatorRetrievalException {
+    return (T) getAggregatorValues(name).getValues().iterator().next();
```
Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResult.java]: `Iterables.getOnlyElement()`?

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 3]
```diff
@@ -479,7 +479,32 @@ public void processWatermark1(Watermark mark) throws Exception {
...
+
+          pushbackDoFnRunner.processElement(elem);
+        }
+      }
+
+      setPushedBackWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+
+      pushbackDoFnRunner.finishBundle();
+
+      // maybe output a new watermark
```
Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Side inputs should never hold the watermark, except so far as delaying elements in the main input until the Side Input is ready

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java [hunk 4]
```diff
@@ -79,6 +86,31 @@ public PipelineResult run(Pipeline pipeline) {
...
     return delegate.getPipelineOptions();
   }
+
+  private void assertAssertionCounters(Pipeline pipeline, FlinkRunnerResult result) {
+    int expectedNumberOfAssertions = PAssert.countAsserts(pipeline);
+    Integer succeededAssertions = 0;
+    try {
+      succeededAssertions = result.getAggregatorValue(PAssert.SUCCESS_COUNTER);
+    } catch (AggregatorRetrievalException e) {
+      // No assertions registered will cause an AggregatorRetrievalException here.
```
Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java]: Why not guard this with an `if` statement on `expectedNumberOfAssertions`?

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java [hunk 5]
```diff
@@ -79,6 +86,31 @@ public PipelineResult run(Pipeline pipeline) {
...
+    }
+    Integer failedAssertions = 0;
+    try {
+      failedAssertions = result.getAggregatorValue(PAssert.FAILURE_COUNTER);
+    } catch (AggregatorRetrievalException e) {
+      // No assertions registered will cause an AggregatorRetrievalException here.
+    }
+    assertThat(
```
Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java]: The documentation on `TestPipeline` says that you have to throw the failing assertion message (https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java#L95)

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResult.java (hunk 2 above)
Comment (aviemzur) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkRunnerResult.java]: 👍

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java (hunk 5 above)
Comment (aviemzur) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java]: This holds true. All tests that assert this pass. For example: https://github.com/apache/beam/blob/d48c0f3f12920f4d2bdd5e6a719cd9bc8f20cc5f/sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java#L427-L427

This assertion is a safety in case we have silent failures.

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java (hunk 4 above)
Comment (aviemzur) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/TestFlinkRunner.java]: 👍

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
Retest this please.

**Comment by aviemzur** [comment]
Retest this please.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java (hunk 1 above)
Comment (aljoscha) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java]: The real check about whether to process main-input elements or not is done here: https://github.com/apache/beam/blob/master/runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java#L83

In reality, if this check here fails it's already too late. Tests were failing because we didn't process main-input data for missing side inputs. This change makes sure that we process pushed back elements when we get the `+Inf` watermark on the side input: https://github.com/apache/beam/pull/2263/files#diff-bb49b4e5c16cd7d4f80a489170f7851aR482. For this to work we needed to remove this check to allow getting an empty side input.

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java (hunk 3 above)
Comment (aljoscha) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: This code runs when we get the `+Inf` watermark on the side input(s). Above, we make sure to flush out all pushed back values.

This made me realise that we have to take into account the watermark hold here. The code should be 
```
// maybe output a new watermark
Instant watermarkHold = stateInternals.watermarkHold();

long potentialOutputWatermark = Math.min(currentInputWatermark, watermarkHold.getMillis());

if (potentialOutputWatermark > currentOutputWatermark) {
  currentOutputWatermark = potentialOutputWatermark;
  output.emitWatermark(new Watermark(currentOutputWatermark));
}```

Comment (aljoscha) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Ah, forget that last comment. `processWatermark1()`, which we were calling there already has this code.

Comment (aljoscha) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: To answer the original question: this code tries to update the output watermark because we evicted a bunch of pushed back values that might have held back the watermark.

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

File: runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java (hunk 3 above)
Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Those elements are processed, though? So the flow is something like
Get elements;
Check side inputs, not ready, push back elements;
Side input watermark goes to infinity
  -> Go read pushed back elements (they don't arrive again, so the watermark processing doesn't happen again by default) -> reprocess the watermark
?
(Really high level, of course)

Comment (aljoscha) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: I think roughly yes, though not all in this method. This method just does:
 - notice that we get `+Inf` watermark on side input -> we will never again see any side input
 - process pushed back values, even though their side input might be empty (some failing tests actually verified that this was the case)
 - update the output watermark. It might still be held back by the watermark hold in case this operator is a `WindowDoFnOperator`. That's why we go through that logic in `processWatermark1()` instead of simply forwarding the `+Inf` watermark.

Comment (tgroh) [inline runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Alright, makes sense.

**Comment by aviemzur** [comment]
Run Flink RunnableOnService

**Comment by aviemzur** [comment]
It looks like `org.apache.beam.sdk.transforms.CombineTest.testSessionsCombineWithContext` is flaky and sometimes passes and sometimes fails on not enough successful assertions.
This could either be an issue with the code it tests, or it could be an issue with the counters themselves that they are overcounting in some cases and that causes us to have incorrect test successes in this and other tests.

**Comment by aviemzur** [comment]
When https://github.com/apache/beam/pull/2417 is merged this PR will no longer be valid as it relies on the aggregators in `PAssert`.
Closing the PR and leaving the branch open for reference, as the other commits by @aljoscha will probably be needed when another solution is found
